seed/seedwriter/seed16.go

```diff
@@ -108,6 +111,35 @@ func (pol *policy16) checkBase(info *snap.Info, availableSnaps *naming.SnapSet)
 	return fmt.Errorf("cannot add snap %q without also adding its base %q explicitly", info.SnapName(), info.Base)
 }
 
+func (pol *policy16) needsImplicitSnaps(availableSnaps *naming.SnapSet) (bool, error) {
+	// do we need to add implicitly either snapd (or core)
+	hasCore := availableSnaps.Contains(naming.Snap("core"))
+	if len(pol.needsCore) != 0 && !hasCore {
+		// XXX warning on Core 18
+		return true, nil
+	}
+
+	if len(pol.needsCore16) != 0 && !hasCore {
+		return false, fmt.Errorf(`cannot use %s requiring base "core16" without adding "core16" (or "core") explicitly`, strutil.Quoted(pol.needsCore16))
+	}
+
+	if pol.model.Classic() && !availableSnaps.Empty() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (pol *policy16) implicitSnaps(availableSnaps *naming.SnapSet) []*asserts.ModelSnap {
+	if len(pol.needsCore) != 0 && !availableSnaps.Contains(naming.Snap("core")) {
```

Comment:
Do we need to look at pol.needsCore16 too?

Reply:
this is https://github.com/snapcore/snapd/blob/master/image/image.go#L485